tutorial/nixify/1_nix_flake.md

```diff
@@ -0,0 +1,223 @@
+Tl;dr Here's the `flake.nix` for this section:
+```nix
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  };
+  outputs = { self, nixpkgs }:
+    let
+      system = "aarch64-darwin";
+      pkgs = nixpkgs.legacyPackages.${system};
+      overlay = final: prev: {
+        todo-app = final.callCabal2nix "todo-app" ./. { };
+      };
+      myHaskellPackages = pkgs.haskellPackages.extend overlay;
+    in
+    {
+      devShells.${system}.default = myHaskellPackages.shellFor {
+          packages = p : [
+            p.todo-app
+          ];
+          buildInputs = with myHaskellPackages; [
+            ghcid
+            cabal-install
+            todo-app
+          ];
+        };
+    };
+}
+```
+
+### shellFor
+
+We will be using `shellFor`, a function defined in `haskellPackages` to set up the default shell for our project. `shellFor` is tailor made for Haskell projects, hence we choose this over `mkShell`. We only need to define two attributes, `packages` and `buildInputs`. `packages` is used to provide the Nix build instructions that we defined earlier for todo-app, so that the shell doesn't start building todo-app again. `buildInputs` is used to ensure that the specified packages are present in the `PATH` of the isolated development environment.
+
+### Let's run!
+
+- `nix develop` this will start a development shell
+- `cabal repl` loads all project modules
+- `ghcid -c "cabal repl exe:todo-app"` reloads your repl session everytime there is a code change
+
+## Nixify external dependencies
+
+Most of real-world applications will rely on external dependencies like Postgres, MySQL, Redis and so on. In this section we will look at how you can start a postgres server using `nix run .#postgres` and also create the data directory local to the project.\
+Tl;dr Here's the `flake.nix`:
+```nix
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  };
+  outputs = { self, nixpkgs }:
+  let
+    system = "aarch64-darwin";
+    pkgs = nixpkgs.legacyPackages.${system};
+  in
+  {
+    apps.${system}.postgres = {
+      type = "app";
+      program = 
+        toString 
+          (pkgs.writeShellApplication {
+            name = "pg_start";
+            runtimeInputs = [ pkgs.postgresql ];
+            text = 
+            ''
+              # Initialize a database with data stored in current project dir
+              [ ! -d "./data/db" ] && initdb --no-locale -D ./data/db
+
+              postgres -D ./data/db -k "$PWD"/data
+            '';
+          }) + "/bin/pg_start";
+    };
+  };
+}
+```
+### writeShellApplication
+
+- A function that generates a derivation to run the shell script in an isolated environment. 
```

Comment:
Does it, in an isolated environment? The script will run any environment you run it in.